### DISCUSSION TITLE
Compatibility with same .proto in different versions

### DISCUSSION THREAD

**Discussion by Alex-abro** [ascopes/protobuf-maven-plugin#415] [Ideas]
For us it is important to be able to use a proto API in different versions at the same time. This means we must be able to compile two .proto files that are completely identically, only the `option java_package` is different. Especially there are same messages in the same proto package.
Example:
`Test.proto` lies in /resources/port/proto/v1
```protobuf
syntax = "proto3";

option java_package = "a.b.c.v1";
option java_outer_classname = "TestProto";
package de.ivu.protobuf;

message A {
...
}
...
```
`Test2.proto` lies in /resources/port/proto/v2
```protobuf
syntax = "proto3";

option java_package = "a.b.c.v2";
option java_outer_classname = "TestProto";
package de.ivu.protobuf;

message A {
...
}
...
```

The messages of both .proto files are identically.

When I compile this with the acopes protobuf-maven-plugin i get
`v2/Test2.proto:15:9: "de.ivu.protobuf.A" is already defined in file "v1/Test.proto".` and
`v2/Test2.proto:19:3: "de.ivu.protobuf.A" seems to be defined in "v1/Test.proto", which is not imported by "v2/Test2.proto".  To use it here, please add the necessary import.`

When using the `com.github.os72` `protoc-jar-maven-plugin` instead with configuration
```xml
<plugin>
  <groupId>com.github.os72</groupId>
  <artifactId>protoc-jar-maven-plugin</artifactId>
  <version>${protoc-jar-maven-plugin.version}</version>
  <executions>
    <execution>
      <phase>generate-test-sources</phase>
      <goals>
        <goal>run</goal>
      </goals>
      <configuration>
        <protocArtifact>com.google.protobuf:protoc:${protobuf.version}</protocArtifact>
        <inputDirectories>
          <include>src/main/proto/resources/port/proto</include>
        </inputDirectories>
      </configuration>
    </execution>
  </executions>
</plugin>
````

it compiles.

I think, the os72 plugin compiles each protobuf file in turn. This makes this feature hard to realize, but maybe you can make some magic :)

**Comment by ascopes** [comment]
Hey, thanks for the message!

The way I set this up to behave like packages in Java was really an intentional design decision at the time, such that it mirrored the general suggested structure for invoking protoc and kept as close to the standard way Maven lays things out as possible. This was one of the reasons that I avoided compiling each file individually at the time. 

I totally appreciate why this is useful as well, but it sounds like something we generally would want protoc to detect since it would usually indicate two sources having the same logical identifier (just like if the Java classes themselves conflicted). I'd usually expect that sort of naming collision to be treated as a bug, since protobuf tries to encourage the use of making messages backwards compatible and avoiding breaking changes on the same API. Likewise if the API version isn't in the message itself or the package, I assume a consumer of the serialized messages would not have a way of knowing which API to use?

Would versioning the protobuf file package itself be a usable workaround in this case? If not, you should be able to create multiple invocations of this plugin per API version, which is clunky but should work. That'd look something like this:

```xml
<executions>
  <execution>
    <id>generate-v1-protobuf-sources</id>
    <phase>generate-sources</phase>
    <goals>
      <goal>generate</goal>
    </goals>
    <configuration>
      <includes>port/proto/v1/**</includes>
    </configuration>
  </execution>

  <execution>
    <id>generate-v2-protobuf-sources</id>
    <phase>generate-sources</phase>
    <goals>
      <goal>generate</goal>
    </goals>
    <configuration>
      <includes>port/proto/v2/**</includes>
    </configuration>
  </execution>
</executions>
```

Another option that I could see is having a separate module for each API version, which would also allow them to be separate concerns, assuming v1 APIs can't be used with v2 APIs in the same messages?

Would any of these be able to work for you, do you think?

**Reply by Alex-abro** [reply]
Thanks for this fast reply!
We can not modify the proto package, because we later need it in the meta information to identify a message.
But I think the second option with the multiple executions works for us. 
I tried it and for me it looks as what we need. I asked my colleagues but we have something like a holiday today so I can't say until Monday.

**Reply by Alex-abro** [reply]
Yeah, it works for us. Thank you!
